### PR TITLE
PAYM-2878 JSON event schema validation

### PR DESCRIPTION
**Pull request by dabeat**
#### :tophat: Problem
Events are not being validated against a JSON schema.

#### :pushpin: Solution
Use [ex json schema](https://github.com/jonasschmidt/ex_json_schema) for JSON schema validation before publishing the event.

#### :ghost: GIF
giphy

### CODE REVIEW

File: lib/events/event_handler.ex [hunk 1]
```diff
@@ -56,6 +61,46 @@ defmodule ElixirTools.Events.EventHandler do
...
+      error_info =
+        event |> Map.from_struct() |> Map.put(:reason, "Event schema validation failed.")
+
+      not_sent_event_module.create!(%{content: Jason.encode!(error_info)})
+    end
+
+    :ok
+  end
+
+  @spec valid_event?(event_schema, Event.t()) :: boolean
```
Comment (dabeat) [inline lib/events/event_handler.ex]: If you agree overall with this PR, I'll move this function to either a dedicated module or to `ElixirTools.Events.Event` (I prefer the second option).

File: lib/events/event_handler.ex [hunk 2]
```diff
@@ -56,6 +61,46 @@ defmodule ElixirTools.Events.EventHandler do
...
+  defp valid_event?(schema, event) do
+    stringified_keys_event = event |> Map.from_struct() |> map_to_string_keys()
+
+    Validator.valid?(schema, stringified_keys_event)
```
Comment (Kitton) [inline lib/events/event_handler.ex]: we can use here `ExJsonSchema.Validator.validate` - it returns some handy error messages

File: lib/events/event_handler.ex [hunk 3]
```diff
@@ -56,6 +61,46 @@ defmodule ElixirTools.Events.EventHandler do
...
+  end
+
+  @spec map_to_string_keys(map) :: map
+  defp map_to_string_keys(map) do
+    Map.new(map, fn {key, value} ->
+      formatted_value =
+        cond do
+          is_map(value) -> map_to_string_keys(value)
```
[the remark this answers is not in the document]
Comment (dabeat) [reply lib/events/event_handler.ex]: Validation crashes.

File: lib/events/event_handler.ex [hunk 4]
```diff
@@ -56,6 +61,46 @@ defmodule ElixirTools.Events.EventHandler do
...
+  @spec valid_event?(event_schema, Event.t()) :: boolean
+  defp valid_event?(schema, event) do
+    stringified_keys_event = event |> Map.from_struct() |> map_to_string_keys()
+
+    Validator.valid?(schema, stringified_keys_event)
+  end
+
+  @spec map_to_string_keys(map) :: map
+  defp map_to_string_keys(map) do
+    Map.new(map, fn {key, value} ->
```
[the remark this answers is not in the document]
Comment (dabeat) [reply lib/events/event_handler.ex]: I modified a solution for a similar purpose I found in a forum.

File: lib/events/event_handler.ex [hunk 5]
```diff
@@ -56,6 +61,46 @@ defmodule ElixirTools.Events.EventHandler do
...
+
+  @spec valid_event?(event_schema, Event.t()) :: boolean
+  defp valid_event?(schema, event) do
+    stringified_keys_event = event |> Map.from_struct() |> map_to_string_keys()
+
+    Validator.valid?(schema, stringified_keys_event)
+  end
+
+  @spec map_to_string_keys(map) :: map
+  defp map_to_string_keys(map) do
```
Comment (Kitton) [inline lib/events/event_handler.ex]: that's what I mentioned in elixir meeting
```elixir
iex(4)> Jason.encode!(%{a: 1}) |> Jason.decode!
%{"a" => 1}

**Pull request closed without merging**